src/main/java/org/truffleruby/core/string/StringNodes.java

```diff
@@ -4167,4 +4169,20 @@ public Rope stringAppend(DynamicObject string, DynamicObject other) {
 
     }
 
+    @Primitive(name = "string_to_null_terminated_byte_array", needsSelf = false)
+    public static abstract class StringToNullTerminatedByteArrayNode extends PrimitiveArrayArgumentsNode {
+
+        @Specialization
+        protected TruffleObject stringToNullTerminatedByteArray(DynamicObject string,
+                @Cached("create()") RopeNodes.BytesNode bytesNode) {
+            // NOTE: we always need one copy here, as native code could modify the passed byte[]
```

Comment:
I had code at one point for a tree to copy its contents into a pre-allocated buffer. I'll look at resurrecting that. Then we could avoid a copy and allocation here.

Reply:
We need one copy for safety anyway, but we could use flattenBytes() with a buffer with one extra 0 at the end instead of getBytes().
It sounds better to flatten in the String too, though, since we might do a `write()` call multiple times on a socket.